### PR TITLE
[IMP] payment_xendit: tokenization

### PR DESCRIPTION
**Pull request by nni-odoo**
Currently, in the APAC region, there are no payment providers
implemented that support tokenization. This commit adds tokenization
support for the `Card` payment method through Xendit. As tokenization
can not be implemented via form redirect, `Card` payments will
specifically be using a direct flow. The changes made here will only
affect `Card` payments, while other payment methods will continue using
redirect implementation.

task-3704600

### CODE REVIEW

**Review by chevalierv — CHANGES_REQUESTED** [review]
Hello @nni-odoo 

Thank you for your work on this PR! I didn't have time to finish the review today so it's only a first part :)

Can you take some time to fix the CI/style issues, and make an upgrade script to remove the view? (See the red runbot) Also, please write a good commit message and document your changes in the `readme.md` file of `payment_xendit`.

Have a nice day!

File: addons/payment_xendit/static/src/js/payment_form.js
Comment (chevalierv) [inline addons/payment_xendit/static/src/js/payment_form.js:107]: `currency` should be passed to Xendit to avoid discrepancies between the platforms. According to [Xendit's documentation](https://developers.xendit.co/api-reference/#create-token), if `currency` is not given, it defaults based on your business country.

File: addons/payment_xendit/models/payment_provider.py
Comment (nni-odoo) [reply addons/payment_xendit/models/payment_provider.py:61]: was following adyen for this one, but indeed there seems to be no use here, will remove it thanks

Comment (nni-odoo) [reply addons/payment_xendit/models/payment_provider.py:61]: yep, will remove this too

File: addons/payment_xendit/static/src/js/payment_form.js
Comment (nni-odoo) [reply addons/payment_xendit/static/src/js/payment_form.js:109]: `tokenization_requested` or any variable that mentions tokenization exist in processingValues, will add it from the `_get_specific_processing_values`

File: addons/payment_xendit/static/src/js/payment_form.js
Comment (nni-odoo) [reply addons/payment_xendit/static/src/js/payment_form.js:113]: putting it on `_prepareInlineForm` after JS is loaded

File: addons/payment_xendit/static/src/js/payment_form.js
Comment (nni-odoo) [reply addons/payment_xendit/static/src/js/payment_form.js:114]: yep will add this flow

**Comment by nni-odoo** [comment]
@chevalierv thank you for the review! made some amendments per your suggestions!

File: addons/payment_xendit/models/payment_transaction.py
Comment (chevalierv) [inline addons/payment_xendit/models/payment_transaction.py:143]: Why do you need to check if a token is not already created? This situation should be impossible as we write the token and `tokenize` at the same time on the transaction. 
```suggestion
            if self.tokenize:
```

File: addons/payment_xendit/models/payment_transaction.py
Comment (chevalierv) [inline addons/payment_xendit/models/payment_transaction.py:1]: In the very unlikely case where `_get_specific_rendering_values` when using the `Card` payment method, an error will be thrown preventing making the operation (validation|payment). I'd suggest adding a check based on the payment method.

https://github.com/odoo/odoo/blob/e069c3a0db3dbcf986d62acac1c9aa3ed36a2a19/addons/payment_xendit/models/payment_transaction.py#L42
```diff
-     if self.provider_code != 'xendit':
+     if self.provider_code != 'xendit' and self.payment_method != 'card':
```

File: addons/payment_xendit/static/src/js/payment_form.js
Comment (chevalierv) [reply addons/payment_xendit/static/src/js/payment_form.js]: nitpicking: I would have done that in `_initiatePaymentFlow` to be able to set a [custom validity](https://developer.mozilla.org/en-US/docs/Web/API/HTMLObjectElement/setCustomValidity)

**Review by chevalierv — CHANGES_REQUESTED** [review]
Hello @nni-odoo,

Here is my second review pass :duck: 

I've pushed some nitpicking changes mainly about docstrings and also, I've changed my mind about the `is_multiple_use` thing so I've reverted the changes myself :) In the end, keeping it like this is cleaner and allows us to avoid adding more things like `_get_specific_processing_values`. I've moved the `currency` params to `_xendit_create_charge` following [my comment](https://github.com/odoo/odoo/pull/158445#discussion_r1539847023).

As I asked you in my previous comment, please write a good commit message and document your changes in the `readme.md` file of `payment_xendit`. Also, don't forget to answer all unresolved comments and leave them unresolved if you answered so I can see it. (you can simply resolve them if you decide to apply the changes :smile:  ).

Have a nice day!

**Comment by nni-odoo** [comment]
Thanks again @chevalierv ! Missed the commit message & README edit message my bad.
I've changed the original commit message and added some more things on the README. Let me know if that's enough.
Other than that,  I've also adapted the implementation according to your suggestions! 
Have  a good day!

**Review by chevalierv — CHANGES_REQUESTED** [review]
Hello @nni-odoo 

Here is my review :smiley: 

Don't forget to squash your commits and apply or answer all comments. For example, see https://github.com/odoo/odoo/pull/158445#discussion_r1539756202.

Regarding your commit message, please focus more on the why than on the what. Here is an example :
```
Currently, in the APAC region, Odoo doesn't support tokenization on any available payment provider in this area. 

This commit adds tokenization support for the `Card` payment method through Xendit. As Xendit does not offer an easy redirect implementation for tokenization, `Card` payments will now be using a direct flow. This does not affect other payment methods supported by Xendit which are still using the redirect implementation.
```

Have a nice day!

File: addons/payment_xendit/static/src/js/payment_form.js
Comment (chevalierv) [reply addons/payment_xendit/static/src/js/payment_form.js:82]: Also, please ensure that those messages will be translated, as any other message in your PR.
```suggestion
            formInputs['card'].setCustomValidity(_t("Invalid Card Number"))
```

File: addons/payment_authorize/models/payment_transaction.py
Comment (nni-odoo) [reply addons/payment_authorize/models/payment_transaction.py:230]: oh right my bad, I edited the wrong file

File: addons/payment_xendit/static/src/js/payment_form.js
Comment (nni-odoo) [inline addons/payment_xendit/static/src/js/payment_form.js:86]: @chevalierv for this one, I think I'll do `_t("Invalid Date")` for both form inputs because suppose date is (10/2022), the invalid one is year and it would be pretty misleading to say month is invalid. Let me know if you have a better sugggestion

Comment (chevalierv) [reply addons/payment_xendit/static/src/js/payment_form.js:86]: LGTM

**Comment by nni-odoo** [comment]
Hello @chevalierv , thank you again for the reviews!
I've made change to commit message and put more details on the README. 
About squashing commits, not sure if you want me to squash it now or later on towards the end of review when code is about to be merged? Let me know if there's anything else to fix

Nice day to you too!

**Comment by chevalierv** [comment]
Hello @nni-odoo 

I'd like you to squash your commits now, there is no need to wait :)

Also, please answer all unresolved comments (I see that https://github.com/odoo/odoo/pull/158445#discussion_r1539756202 is still unresolved)

**Comment by nni-odoo** [comment]
Hello @chevalierv squashed all the commits and resolved all the unanswered comments!

File: addons/payment_xendit/static/src/js/payment_form.js
Comment (chevalierv) [inline addons/payment_xendit/static/src/js/payment_form.js:176]:
```suggestion
```

File: addons/payment_xendit/static/src/js/payment_form.js
Comment (chevalierv) [inline addons/payment_xendit/static/src/js/payment_form.js:194]:
```suggestion
```

**Review by chevalierv — APPROVED** [review]
Still two minors comments :) Please fix it straight away (I mean don't wait for the final review)

Otherwise it LGTM :+1: 

Thanks for your collaboration :smiley:

File: addons/payment_xendit/models/payment_transaction.py
Comment (chevalierv) [inline addons/payment_xendit/models/payment_transaction.py:30]: Your runbot is red :eyes: 
```suggestion
        if self.provider_code != 'xendit' and self.payment_method_code != 'card':
```

**Comment by nni-odoo** [comment]
@chevalierv fixed the issues and I've re-tested the whole flow again, seems to be smooth from my side.

Thank you again for your help on this!!

File: addons/payment_xendit/models/payment_transaction.py
Comment (chevalierv) [inline addons/payment_xendit/models/payment_transaction.py:125]: runbot is red :eyes: 
Why did you add this? We need to process all transactions regardless of the payment method.
```suggestion
        if self.provider_code != 'xendit':
```

**Comment by nni-odoo** [comment]
@chevalierv my bad on that one. I pushed again the fixes but it seems there's still red on runbot. When I look into it, the error seems to not be correlated to the changes introduced?

**Comment by chevalierv** [comment]
Hello @nni-odoo 

On this specific PR, we expected to see a red runbot for CI/template and CI/security. We'll take care of it we'll do the final review. 

For the ci/runbot, I just checked and IMO, it's not related to your changes. You could try to rebase your branch to see if the issues has been fixed.

**Comment by nni-odoo** [comment]
@chevalierv rebased and runbot is green now! thanks!

File: addons/payment_xendit/controllers/main.py
Comment (AntoineVDV) [inline addons/payment_xendit/controllers/main.py]: I'm a bit concerned by the security aspect of this call. Basically, if an attacker finds someone else's token ID, they can use it to pay for their own transaction that was created beforehand.

File: addons/payment_xendit/controllers/main.py
Comment (AntoineVDV) [inline addons/payment_xendit/controllers/main.py]: `partner_id` is never used.

File: addons/payment_xendit/controllers/main.py
Comment (AntoineVDV) [inline addons/payment_xendit/controllers/main.py]: Shouldn't you rather create the token in Odoo and set Xendit's token_id as `provider_ref`, then call `_send_payment_request`?

File: addons/payment_xendit/models/payment_transaction.py
Comment (AntoineVDV) [inline addons/payment_xendit/models/payment_transaction.py:193]: What do you mean by "...Xendit objects. See _process_notification_data"?

**Review by AntoineVDV — CHANGES_REQUESTED** [review]
Hi, sorry for the late review :innocent: 

I applied some changes directly in a fixup commit (after rebasing on master) to save time with the review, but I still had some remarks/questions that I posted here below.

Also, I'm getting this error when trying to process a validation transaction (through /my/payment_method)
image
I might have broken something without noticing... could you check? :pray: 

Finally, could you make sure that the test coverage is complete?

File: addons/payment_xendit/static/src/js/payment_form.js
Comment (nni-odoo) [reply addons/payment_xendit/static/src/js/payment_form.js:120]: It will not be allowed, I believe an error would be thrown right away while trying to create token (which I believe is already handled currently), so I suppose no adjustmends needed here?

File: addons/payment_xendit/static/src/js/payment_form.js
Comment (nni-odoo) [reply addons/payment_xendit/static/src/js/payment_form.js]: I'll make the input required in this case

File: addons/payment_xendit/controllers/main.py
Comment (nni-odoo) [reply addons/payment_xendit/controllers/main.py]: `_xendit_create_charge` calls `_xendit_make_request`  which will still require the `xendit_secret_key` field when calling the API . So I think that attacker would only be successful when he knows both the token and also the secret key. 
If the secret key is already exposed, I think attacker would pretty much be able to do anything :laughing:

Comment (nni-odoo) [reply addons/payment_xendit/controllers/main.py]: this is true. The reason I do this is, when I was developing it, the token created turns out to be invalid (because it was not multiple-use), which in this case results to unsuccessful payment, however the token is still created somehow.
So I'm doing this for safety reason to make sure that payment is successful first before storing the token in Odoo. Let me know what you think and if we should still proceed with changing the logic here

File: addons/payment_xendit/static/src/js/payment_form.js
Comment (AntoineVDV) [reply addons/payment_xendit/static/src/js/payment_form.js:120]: No need, indeed; I just wanted confirmation ;)

File: addons/payment_xendit/controllers/main.py
Comment (AntoineVDV) [reply addons/payment_xendit/controllers/main.py]: They can call the route with an arbitrary reference and token ID, so they would actually only need to know the latter. I will check with the security team.

Comment (nni-odoo) [reply addons/payment_xendit/controllers/main.py]: that sounds like a good idea, will do that thanks!

Comment (nni-odoo) [reply addons/payment_xendit/controllers/main.py]: Hello @AntoineVDV for this one, I tried implementing it and have some concerns with the approach mentioned:

* we no longer have this _processPayment() function so I think it's not possible to model it after this SEPA one. 
* The possible option that I see is overriding the `_initiatePaymentFlow()`, create token, add it to paymentContext then update the flow to token. However I think with this approach, this means that we're forcing it to always save the payment method even if they don't tick that option when entering card details. 
* Payment processing cannot proceed with an archived token, so I don't think we can make it archived first.
* If we override rom the `_processDirectFlow`, the transactions is already created, not sure if it's a good idea to attach the token_id to transaction from this point on?

For these points, I believe it's best to keep the current implementation as is. Let me know if you think there's something better to do here. I checked all other payment methods now including SEPA and I don't think this way is currently implemented anymore...

**Comment by nni-odoo** [comment]
@AntoineVDV made the changes on PR per your review , as for the error you've screenshoted, I think I encountered before based on the error, and I believe it should be related to the credit card payment configuration? (try this URL after logging in: https://dashboard.xendit.co/settings/payment-methods/cards-configuration)
image

File: addons/payment_xendit/controllers/main.py
Comment (AntoineVDV) [reply addons/payment_xendit/controllers/main.py]: > - Payment processing cannot proceed with an archived token, so I don't think we can make it archived first.

That was my main concern and reason not to go this road.

Let's keep the current solution then ;)

> - we no longer have this _processPayment() function so I think it's not possible to model it after this SEPA one.
> - The possible option that I see is overriding the _initiatePaymentFlow(), create token, add it to paymentContext then update the flow to token. However I think with this approach, this means that we're forcing it to always save the payment method even if they don't tick that option when entering card details.
> - If we override rom the _processDirectFlow, the transaction is already created, not sure if it's a good idea to attach the token_id to transaction from this point on? If we do the _initiatePaymentFlow , this means we will create a new transaction again.

We have equivalent methods in 17.0 that could have met our requirements but no need to bother anymore if we create the token after payment.

Comment (nni-odoo) [reply addons/payment_xendit/controllers/main.py]: Noted with thanks! will resolve the convo in that case

**Comment by nni-odoo** [comment]
@AntoineVDV noted on that, I've made the PR for docu update here: https://github.com/odoo/documentation/pull/9394

Thank you again!

File: addons/payment_xendit/controllers/main.py
Comment (AntoineVDV) [reply addons/payment_xendit/controllers/main.py]: @odoo/rd-security could we have your input here?

**Review by xmo-odoo — CHANGES_REQUESTED** [review]

**Comment by AntoineVDV** [comment]
Hello @xmo-odoo could you please check https://github.com/odoo/odoo/pull/158445#discussion_r1604595805? :pray:

File: addons/payment_xendit/controllers/main.py
Comment (nni-odoo) [reply addons/payment_xendit/controllers/main.py]: @xmo-odoo thank you for your input here! Indeed, Xendit allows a single-use token as well (make the is_multiple_use: False)

In this case, I've made the change such that single/multi-use is conditional on whether user wants to save the payment method. That's what I'm getting from your feedback, but  please do correct me if I'm wrong or if there's still anything I miss!
cc @AntoineVDV

**Comment by nni-odoo** [comment]
@xmo-odoo 
for `should_authenticate`, it's actually for convenience purposes so that it won't ask for verification, I'll look into the flow if we let `should_authenticate` to true.

To my understanding from reading the code, isn't the `paymentContext.tokenizationRequested` related to whether or not user checks the "Save My Payment Details" when paying with credit card, which then is passed when creating transaction (field: tokenize). This is where I put the condition of `self.tokenize`, which will create the token in Odoo and save it to the transaction too. So with current implementation, single-use tokens won't be saved but multi-use tokens will.

Hope this addresses your concern and let me know if I missed anything here,
thank you again for reviewing!

**Comment by nni-odoo** [comment]
@xmo-odoo @AntoineVDV To follow up on earlier statement regarding the `should_authenticate=False`. If it's true, it will opt for user verification as such 
image

I'm wondering if we've implemented something similar before? If so, any reference I could use?
On the other hand if not, is it something you think is worth exploring on? 

Let me know if you have further questions!

**Comment by nni-odoo** [comment]
@AntoineVDV adding little context to the latest commit

since we're adding authentication flow, I'm taking reference from xendit themselves:
- https://docs.xendit.co/credit-cards/integrations/tokenization (reference)
- https://js.xendit.co/test_tokenize.html (sample form)

I'm adding `auth_ui.js` and `auth_service.js` that follows how `block_ui` is applied. Since we need to still have the BlockUI (to prevent user choosing other payment methods), then we'll have to create new component "above" BlockUI. I came up with adding z-index but z-index will only work with blocks at the same level (ref: https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_positioned_layout/Understanding_z-index/Stacking_context). And that's why I have to put it into the `main_component`. 

Let me know if you have other suggestions that could simplify this, thanks!

**Review by AntoineVDV — APPROVED** [review]
Hello @nni-odoo, I think we're good to go! :rocket: 

TBH, I have doubts about this new auth_ui/service thing because, as you said, it's a bit "hacky-hacky". There might be a simpler way to achieve the same result, but since it's only frontend assets (it can be changed easily), does work, and is not worth spending too much time on, let's keep it as-is ;)

**Comment by AntoineVDV** [comment]
@xmo-odoo Could you please add the ignore ci/security when you consider all security concerns have been answered?

**Comment by AntoineVDV** [comment]
@robodoo r+
@robodoo override=ci/security
Reason: `window.location`: Redirection to a fixed, hard-coded, relative URL.
Link to ci/security build (master): https://runbot.odoo.com/runbot/build/64109175

**Pull request closed without merging**
